src/services/persistent-data/persistent-data.service.ts

```diff
@@ -0,0 +1,48 @@
+import {
+    DOCUMENT
+} from '@angular/platform-browser';
+import {
+    Injectable,
+    Inject
+} from '@angular/core';
+
+export class PersistentDataService {
+
+    setData(key: string, value: string, storageType?: PersistentDataStorageType) {
+        if (!storageType && localStorage || storageType === PersistentDataStorageType.LocalStorage) {
+            localStorage.setItem(key, value);
+        } else if (!storageType || storageType !== PersistentDataStorageType.SessionStorage) {
+            document.cookie = key + '=' + value + '; path=/';
+        } else {
+            sessionStorage.setItem(key, value);
+        }
+    }
+
+    getData(key: string, storageType?: PersistentDataStorageType) {
+        if (!storageType && localStorage || storageType === PersistentDataStorageType.LocalStorage) {
+            return localStorage.getItem(key);
+        } else if (!storageType || storageType !== PersistentDataStorageType.SessionStorage) {
+            var keyName = key + '=';
+            var allCookies = document.cookie.split(';');
+            for (var i = 0, max = allCookies.length; i < max; i++) {
+                var cookie = allCookies[i];
+                while (cookie.charAt(0) === ' ') {
+                    cookie = cookie.substring(1, cookie.length);
+                } 
+                if (cookie.indexOf(keyName) === 0) {
+                    return cookie.substring(keyName.length, cookie.length);
+                } 
+            }
+            return null;
+        } else {
+            return sessionStorage.getItem(key);
+        }
+    }
+
+}
+
+export enum PersistentDataStorageType {
```

Comment:
Currently this class isn't very extensible, for example we may in future want to add additional storage options (such as IndexedDB or FileAPI support) which could prove a bit tricky with all the if statements etc.. We also should be using a try..catch to most accurately determine whether or not local/session storage works or not.

Consider the following code which should be a bit more extensible if we want to add more storage types in future:

```typescript
import { Injectable } from '@angular/core';

@Injectable()
export class PersistentDataService {

    /**
     * Save the item in some form of persistent storage
     */
    setItem(key: string, value: string, storageType: PersistentDataStorageType = PersistentDataStorageType.LocalStorage): void {

        // check if storage is supported - otherwise fallback
        storageType = this.getStorageType(storageType);

        // peform the save on the desire storage target
        switch (storageType) {

            case PersistentDataStorageType.LocalStorage:
                this.setItemStorage(localStorage, key, value);
                break;

            case PersistentDataStorageType.SessionStorage:
                this.setItemStorage(sessionStorage, key, value);
                break;

            case PersistentDataStorageType.Cookie:
                this.setItemCookie(key, value);
                break;
        }
    }

    /**
     * Get a stored value from persistent storage
     */
    getItem(key: string, storageType: PersistentDataStorageType = PersistentDataStorageType.LocalStorage): string {

        // get the supported storage type
        storageType = this.getStorageType(storageType);

        switch (storageType) {

            case PersistentDataStorageType.LocalStorage:
                return this.getItemStorage(localStorage, key);

            case PersistentDataStorageType.SessionStorage:
                return this.getItemStorage(sessionStorage, key);

            case PersistentDataStorageType.Cookie:
                return this.getItemCookie(key);
        }
    }

    /**
     * Store value in either local or session storage
     */
    private setItemStorage(storage: Storage, key: string, value: string): void {
        storage.setItem(key, value);
    }

    /**
     * Save the value in a cookie 
     */
    private setItemCookie(key: string, value: string): void {
        document.cookie = key + '=' + value + '; path=/';
    }

    /**
     * Retrieve value in either local or session storage
     */
    private getItemStorage(storage: Storage, key: string): string {
        return storage.getItem(key);
    }

    /**
     * Retrieve the value in a cookie 
     */
    private getItemCookie(key: string): string {

        // get all the cookies for this site
        const cookies = document.cookie.split(';');

        // process the cookies into a from we can easily manage
        const match = cookies
            .map(cookie => ({ key: cookie.split('=')[0].trim(), value: cookie.split('=')[1].trim() }))
            .find(cookie => cookie.key === key);

        return match ? match.value : null;
    }

    /**
     * Determine which storage method can be used by the browser
     */
    private getStorageType(desiredType: PersistentDataStorageType): PersistentDataStorageType {

        // If local storage is desired then make sure we can use it - or fallback to cookies
        if (desiredType === PersistentDataStorageType.LocalStorage) {
            return this.getStorageSupported(localStorage) ? PersistentDataStorageType.LocalStorage : PersistentDataStorageType.Cookie;
        }

        // If session storage is desired then make sure we can use it - or fallback to cookies
        if (desiredType === PersistentDataStorageType.SessionStorage) {
            return this.getStorageSupported(sessionStorage) ? PersistentDataStorageType.LocalStorage : PersistentDataStorageType.Cookie;
        }

        return desiredType;
    }

    /**
     * Perform a test to see if the storage option is supported
     * @param storage Either local storage or session storage
     */
    private getStorageSupported(storage: Storage): boolean {

        // if the storage type is null or undefined then fall back to cookies
        if (!storage) {
            return false;
        }

        try {
            storage.setItem('ux-persistent-data-service', 'ux-persistent-data-service');
            storage.removeItem('ux-persistent-data-service');
            return true;
        } catch (err) {
            return false;
        }
    }

}

export enum PersistentDataStorageType {
    LocalStorage,
    Cookie,
    SessionStorage
}
```

> Note, I havent tested this so you would need to check that it works ok